tests/test_terra_notebook_utils.py

```diff
@@ -86,5 +90,49 @@ def test_extract(self):
                 self.assertGreater(time.time() - start_time, age)
 
 
+class TestTerraNotebookUtilsProgress(unittest.TestCase):
+    def test_progress_reporter(self):
+        with progress.ProgressReporter() as pr:
+            pr.checkpoint(2)
+            pr.checkpoint(5)
+            pr.checkpoint(7)
+            time.sleep(2)
+            self.assertEqual(14, pr.units_processed)
+
+    def test_rate_limited(self):
+        rate = 50
+        with self.subTest("Should be rate limited to expected number of calls"):
+            call_info = dict(number_of_calls=0)
+
+            @progress.RateLimited(rate)
+            def rate_limited_func():
+                call_info['number_of_calls'] += 1
+
+            start_time = time.time()
+            duration = 0
+            while duration <= 4.5 * 1 / rate:
+                rate_limited_func()
+                duration = time.time() - start_time
+            self.assertEqual(5, call_info['number_of_calls'])
+
+        with self.subTest("Should get expected exception for calling too quickly"):
+            class MockRateLimitedException(Exception):
+                pass
+
+            @progress.RateLimited(rate, MockRateLimitedException)
+            def raising_rate_limited_func():
+                pass
+
+            with self.assertRaises(MockRateLimitedException):
+                raising_rate_limited_func()
```

Comment:
Why does this call `raising_rate_limited_func()` twice?

Reply:
Calling it once wouldn't cause the rate limiting logic to trigger.